### PR TITLE
[Memory optm] loss using torch + compile

### PR DESCRIPTION
**Pull request by felipemello1**
Memory freebies
image

------
I dont think that loss/reward is a good way to check correctness here. But i compared the functions locally and they provide the same output.

image

image

### CODE REVIEW

File: src/forge/util/ops.py [hunk 1]
```diff
@@ -65,10 +66,23 @@ def compute_logprobs(
...
     """
     # Ignore the last token from logits because it predicts the next token (-1)
     # And align logits with the input tokens length.
     logits = logits[:, -input_ids.size(1) - 1 : -1, :].to(input_ids.device)
     scaled_logits = logits / temperature
-    logprobs = selective_log_softmax(scaled_logits, input_ids)
-    return logprobs
+
+    # Convert to fp32 for numerical stability
+    scaled_logits_fp32 = scaled_logits.float()
```
Comment (JenniferWang) [inline src/forge/util/ops.py:78]: Noob question: what's the dtype for `scaled_logits`?

Comment (felipemello1) [reply src/forge/util/ops.py:78]: float becomes torch.float32

**Comment by felipemello1** [comment]
@ebsmothers @Jack-Khuu @joecummings @pbontrager can some of you confirm that i dont need to do the all_gather that was happening in `selective_log_softmax`? Maybe its necessary if `trainer.parallelism.disable_loss_parallel=False`? But this is off for all of our configs.

File: src/forge/util/ops.py [hunk 2]
```diff
@@ -65,10 +66,23 @@ def compute_logprobs(
...
+        logprobs: [batch, seq_len] log probabilities for each token
     """
     # Ignore the last token from logits because it predicts the next token (-1)
     # And align logits with the input tokens length.
     logits = logits[:, -input_ids.size(1) - 1 : -1, :].to(input_ids.device)
     scaled_logits = logits / temperature
-    logprobs = selective_log_softmax(scaled_logits, input_ids)
-    return logprobs
+
+    # Convert to fp32 for numerical stability
```
Comment (JenniferWang) [inline src/forge/util/ops.py]:
```suggestion
    # Cast up to fp32 for numerical stability
```

File: src/forge/util/ops.py
Comment (JenniferWang) [inline src/forge/util/ops.py:11]: Should we also delete this function?

**Comment by casteryh** [comment]
<del> just curious, if we do torch.compile on the textbook implementation, do we still need manual optimization like `selective_log_softmax`? </del>
never mind, just  remove the `selective_log_softmax` all together.

File: src/forge/actors/trainer.py [hunk 3]
```diff
@@ -160,6 +160,10 @@ def __post_init__(self):
         }
         os.environ.update(env)
 
+        # compile loss
+        logger.info("Compiling loss")
+        self.loss = torch.compile(self.loss)
```
Comment (joecummings) [inline src/forge/actors/trainer.py:165]: Is there any circumstance under which this command would fail?

File: src/forge/util/ops.py [hunk 4]
```diff
@@ -65,10 +66,23 @@ def compute_logprobs(
...
     scaled_logits = logits / temperature
-    logprobs = selective_log_softmax(scaled_logits, input_ids)
-    return logprobs
+
+    # Cast up to fp32 for numerical stability
+    scaled_logits_fp32 = scaled_logits.float()
+
+    # get per-token log probs
+    batch_size, seq_len, vocab_size = scaled_logits_fp32.shape
+    log_probs = -F.cross_entropy(
```
Comment (joecummings) [inline src/forge/util/ops.py]:
```suggestion
    logprobs = -F.cross_entropy(
```

I will die on this hill.

File: src/forge/util/ops.py [hunk 5]
```diff
@@ -65,10 +66,23 @@ def compute_logprobs(
...
+        logprobs: [batch, seq_len] log probabilities for each token
     """
     # Ignore the last token from logits because it predicts the next token (-1)
     # And align logits with the input tokens length.
     logits = logits[:, -input_ids.size(1) - 1 : -1, :].to(input_ids.device)
     scaled_logits = logits / temperature
-    logprobs = selective_log_softmax(scaled_logits, input_ids)
-    return logprobs
+
+    # Cast up to fp32 for numerical stability
```
Comment (joecummings) [inline src/forge/util/ops.py:77]: nit: I would change this to something like "ensure logits are in fp32" b/c they actually could already be in fp32 and no need for "Casting up"

File: src/forge/actors/trainer.py (hunk 3 above)
Comment (felipemello1) [reply src/forge/actors/trainer.py:165]: cant think of one in our scenario, but if/when this happens, we can fix it

File: src/forge/util/ops.py
Comment (felipemello1) [reply src/forge/util/ops.py:11]: its used in 3 other places. I will leave it there for now. We will prob need some larger refactoring later to clean up / organize losses
